fleetspeak/src/server/cpsservice/service_test.go

```diff
@@ -0,0 +1,137 @@
+// Copyright 2017 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cpsservice
+
+import (
+	"context"
+	"os"
+	"testing"
+	"time"
+
+	"cloud.google.com/go/pubsub"
+	"github.com/google/fleetspeak/fleetspeak/src/common/anypbtest"
+	"github.com/google/fleetspeak/fleetspeak/src/server/service"
+	"google.golang.org/protobuf/proto"
+
+	fspb "github.com/google/fleetspeak/fleetspeak/src/common/proto/fleetspeak"
+	cpspb "github.com/google/fleetspeak/fleetspeak/src/server/cpsservice/proto/fleetspeak_cpsservice"
+	spb "github.com/google/fleetspeak/fleetspeak/src/server/proto/fleetspeak_server"
+)
+
+func requireEmulation(t *testing.T) {
+	if os.Getenv("PUBSUB_EMULATOR_HOST") == "" {
+		t.Skipf("Cloud Pub/Sub emulator not found")
+	}
+}
+
+type testEnv struct {
+	svc service.Service
+	sub *pubsub.Subscription
+}
+
+func createTestEnv(t *testing.T) testEnv {
```

Comment:
If a test helper calls t.Fatalf() or passes the t object onwards, it should call t.Helper() at the top. That makes sure that the failure message is attributed to the surrounding test, rather than to the line within the helper function.

See https://google.github.io/styleguide/go/decisions#must-functions (second example) or https://blog.gnoack.org/post/go-tests/ for some examples.

Reply:
Done.